cmd/server/apiserver.go

```diff
@@ -215,12 +222,117 @@ func (server *APIServer) Start(router *mux.Router) {
 			),
 		)
 	})
+
+	prepareTestSchedule := func(w http.ResponseWriter, r *http.Request) *m.TestSchedule {
+		bodyContent, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			w.Write(buildFailure(err.Error(), http.StatusBadRequest, w))
+			return nil
+		}
+
+		err = m.Validate(reflect.TypeOf(m.TestSchedule{}), bodyContent)
+		if err != nil {
+			w.Write(buildFailure(err.Error(), http.StatusBadRequest, w))
+			return nil
+		}
+
+		var schedule m.TestSchedule
+		err = json.Unmarshal(bodyContent, &schedule)
+		if err != nil {
+			w.Write(buildFailure(err.Error(), http.StatusBadRequest, w))
+			return nil
+		}
+
+		// make sure cron spec is valid
+		err = server.sm.ValidateSpec(schedule.CronSpec)
+		if err != nil {
+			w.Write(buildFailure(err.Error(), http.StatusBadRequest, w))
+			return nil
+		}
+
+		// make sure specified Test exists
+		if _, err := sto.GetTestByTestId(schedule.TestID); err != nil {
+			w.Write(buildFailure(
+				fmt.Sprintf("Cannot find Test<%s>", schedule.TestID),
+				http.StatusNotFound,
+				w,
+			))
+			return nil
+		}
+
+		schedule.ID = m.TestScheduleID(schedule.Name)
+
+		return &schedule
+	}
+
+	router.HandleFunc("/test-schedules", func(w http.ResponseWriter, r *http.Request) {
+		schedule := prepareTestSchedule(w, r)
+		if schedule == nil {
+			return
+		}
+
+		server.sm.Add(schedule)
+
+		w.Write(
+			buildSuccess(
+				map[string]string{
+					"scheduleid": string(schedule.ID),
+				},
+				w,
+			),
+		)
+	}).Methods(http.MethodPost)
+
+	router.HandleFunc("/test-schedules/{id}", func(w http.ResponseWriter, r *http.Request) {
+		vars := mux.Vars(r)
+		scheduleid := vars["id"]
+
+		switch r.Method {
+		case http.MethodGet:
+			// TODO: get from storage
+			w.Write(
+				buildSuccess(
+					map[string]interface{}{},
+					w,
+				),
+			)
+		case http.MethodPut:
+			// TODO: check if TestSchedule actually exists
+
+			schedule := prepareTestSchedule(w, r)
+			if schedule == nil {
+				return
+			}
+
+			// make sure id isn't changed
+			schedule.ID = m.TestScheduleID(scheduleid)
```

Comment:
Q: What to do when the `Name` is changed during an update? The ID remains the original name during creation.

Same Q for Tests

Reply:
are our ID fields still = name + some random string ?

Reply:
no they're just the same as name